File: src/main/java/com/vzome/core/editor/TransformationTool.java

//(c) Copyright 2008, Scott Vorthmann.  All rights reserved.

package com.vzome.core.editor;

import org.w3c.dom.Element;

import com.vzome.core.commands.XmlSaveFormat;
import com.vzome.core.commands.Command.Failure;
import com.vzome.core.construction.Construction;
import com.vzome.core.construction.Point;
import com.vzome.core.construction.Polygon;
import com.vzome.core.construction.Segment;
import com.vzome.core.construction.Transformation;
import com.vzome.core.construction.TransformedPoint;
import com.vzome.core.construction.TransformedPolygon;
import com.vzome.core.construction.TransformedSegment;
import com.vzome.core.model.Manifestation;
import com.vzome.core.model.RealizedModel;

public abstract class TransformationTool extends ChangeManifestations implements Tool
{
    @Override
    public void prepare( ChangeManifestations applyTool ) {}

    @Override
	public void complete( ChangeManifestations applyTool ) {}

    @Override
	public boolean isSticky()
    {
        return true;
    }
    
    @Override
    public boolean needsInput()
    {
    	return true;
    }

    @Override
    public void perform() throws Failure
    {
        defineTool();
    }

    protected void defineTool()
    {
        tools .addTool( this );
    }

    private String name;
    
    protected Transformation[] transforms;
    
    private final Tool.Registry tools;
    
    protected Point originPoint;

    public TransformationTool( String name, Selection selection, RealizedModel realized, Tool.Registry tools, Point originPoint )
    {
        super( selection, realized, false );
        
        if ( name != null && name .endsWith( DEFAULT_NAME_TRIGGER ) )
            this .name = name .replaceAll( DEFAULT_NAME_TRIGGER, getDefaultName( name ) );
        else
            this.name = name;
        this.tools = tools;
        this.originPoint = originPoint;
    }
    
    public static String DEFAULT_NAME_TRIGGER = "____DEFAULT";
    
    protected abstract String getDefaultName( String template );
    
    protected boolean isAutomatic()
    {
        int dot = this .name .indexOf( "." );
        int slash = this .name .indexOf( "/" );
        String id = this .name .substring( dot, slash );
        return ".auto" .equals( id );
    }

    @Override
    public void performEdit( Construction c, ChangeManifestations applyTool )
    {
        for (Transformation transform : transforms) {
            Construction result = null;
            if (c instanceof Point) {
                result = new TransformedPoint(transform, (Point) c);
            } else if (c instanceof Segment) {
                result = new TransformedSegment(transform, (Segment) c);
            } else if (c instanceof Polygon) {
                result = new TransformedPolygon(transform, (Polygon) c);
            } else {
                // TODO handle other constructions 
            }
            if ( result == null )
                continue;
            applyTool .manifestConstruction( result );
        }
        applyTool .redo();
    }
    
    @Override
	public void performSelect( Manifestation man, ChangeManifestations applyTool ) {};

    @Override
    public void redo()
    {
        // TODO manifest a symmetry construction... that is why this class extends ChangeConstructions
        // this edit is now sticky (not really undoable)
//        tools .addTool( this );
    }

    @Override
    public void undo()
    {
        // this edit is now sticky (not really undoable)
//        tools .removeTool( this );
    }

    @Override
    public String getName()
    {
        return name;
    }
    
    @Override
    protected void getXmlAttributes( Element element )
    {
        element .setAttribute( "name", this.name );
    }

    @Override
    protected void setXmlAttributes( Element element, XmlSaveFormat format ) throws Failure
    {
        this.name = element .getAttribute( "name" );
    }
}

File: online/src/worker/legacy/ts/com/vzome/core/tools/TransformationTool.ts
import { java, javaemul } from "../../../../../candies/j4ts-2.1.0-SNAPSHOT/bundle.js";
import { Color } from "../construction/Color.js";
import { Construction } from "../construction/Construction.js";
import { Point } from "../construction/Point.js";
import { Transformation } from "../construction/Transformation.js";
import { Tool } from "../editor/Tool.js";
import { ToolsModel } from "../editor/ToolsModel.js";
import { ChangeManifestations } from "../editor/api/ChangeManifestations.js";
import { Manifestation } from "../model/Manifestation.js";

export abstract class TransformationTool extends Tool {
    /**
     * 
     * @param {ChangeManifestations} applyTool
     */
    public prepare(applyTool: ChangeManifestations) {
    }

    /**
     * 
     * @param {ChangeManifestations} applyTool
     */
    public complete(applyTool: ChangeManifestations) {
    }

    /**
     * 
     * @return {boolean}
     */
    public needsInput(): boolean {
        return true;
    }

    transforms: Transformation[];

    originPoint: Point;

    public constructor(id: string, tools: ToolsModel) {
        super(id, tools);
        if (this.transforms === undefined) { this.transforms = null; }
        if (this.originPoint === undefined) { this.originPoint = null; }
        this.originPoint = tools.getOriginPoint();
    }

    /**
     * 
     * @param {*} that
     * @return {boolean}
     */
    public equals(that: any): boolean {
        if (this === that){
            return true;
        }
        if (that == null){
            return false;
        }
        if (!/* equals */(<any>((o1: any, o2: any) => { if (o1 && o1.equals) { return o1.equals(o2); } else { return o1 === o2; } })((<any>that.constructor),(<any>this.constructor)))){
            return false;
        }
        const other: TransformationTool = <TransformationTool>that;
        if (this.originPoint == null){
            if (other.originPoint != null){
                return false;
            }
        } else if (!/* equals */(<any>((o1: any, o2: any) => { if (o1 && o1.equals) { return o1.equals(o2); } else { return o1 === o2; } })(this.originPoint,other.originPoint))){
            return false;
        }
        if (!java.util.Arrays.equals(this.transforms, other.transforms)){
            return false;
        }
        return true;
    }

    /**
     * 
     * @param {Construction} c
     * @param {ChangeManifestations} applyTool
     */
    public performEdit(c: Construction, applyTool: ChangeManifestations) {
        for(let index = 0; index < this.transforms.length; index++) {
            let transform = this.transforms[index];
            {
                const result: Construction = transform.transform$com_vzome_core_construction_Construction(c);
                if (result == null)continue;
                const color: Color = c.getColor();
                result.setColor(color);
                const m: Manifestation = applyTool.manifestConstruction(result);
                if (m != null)if (color != null)applyTool.colorManifestation(m, c.getColor());
            }
        }
        applyTool.redo();
    }

    /**
     * 
     * @param {*} man
     * @param {ChangeManifestations} applyTool
     */
    public performSelect(man: Manifestation, applyTool: ChangeManifestations) {
    }

    public unselect$com_vzome_core_model_Manifestation$boolean(man: Manifestation, ignoreGroups: boolean) {
        const c: Construction = man.getFirstConstruction();
        this.addParameter(c);
        super.unselect$com_vzome_core_model_Manifestation$boolean(man, ignoreGroups);
    }

    /**
     * 
     * @param {*} man
     * @param {boolean} ignoreGroups
     */
    public unselect(man?: any, ignoreGroups?: any) {
        if (((man != null && (man.constructor != null && man.constructor["__interfaces"] != null && man.constructor["__interfaces"].indexOf("com.vzome.core.model.Manifestation") >= 0)) || man === null) && ((typeof ignoreGroups === 'boolean') || ignoreGroups === null)) {
            return <any>this.unselect$com_vzome_core_model_Manifestation$boolean(man, ignoreGroups);
        } else if (((man != null && (man.constructor != null && man.constructor["__interfaces"] != null && man.constructor["__interfaces"].indexOf("com.vzome.core.model.Manifestation") >= 0)) || man === null) && ignoreGroups === undefined) {
            return <any>this.unselect$com_vzome_core_model_Manifestation(man);
        } else throw new Error('invalid overload');
    }

    isAutomatic(): boolean {
        return /* contains */(this.getId().indexOf(".auto/") != -1);
    }
}
TransformationTool["__class"] = "com.vzome.core.tools.TransformationTool";
TransformationTool["__interfaces"] = ["com.vzome.api.Tool"];
